From 85a5606f3d560c8a7cd4a750eea8d0cf10e35a6d Mon Sep 17 00:00:00 2001
From: Joe Mason <joenotcharles@google.com>
Date: Tue, 26 Mar 2024 16:24:12 +0000
Subject: [PATCH] Handle missing browser context in IsPageOptedOutOfDiscarding

Fixes a crash when CanDiscard is called on a page before the prefs are
fully loaded for its profile.

Bug: 327684692
Change-Id: I81707b153982c3e5e9b4e9e72312e9a0c1201173
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5392267
Commit-Queue: Joe Mason <joenotcharles@google.com>
Reviewed-by: Vovo Yang <vovoy@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1278371}
---

--- a/chrome/browser/performance_manager/policies/page_discarding_helper.cc
+++ b/chrome/browser/performance_manager/policies/page_discarding_helper.cc
@@ -453,10 +453,12 @@
     const std::string& browser_context_id,
     const GURL& url) const {
   auto it = profiles_no_discard_patterns_.find(browser_context_id);
-  // TODO(crbug.com/1308741): Change the CHECK to a DCHECK in Sept 2022, after
-  // verifying that there are no crash reports.
-  CHECK(it != profiles_no_discard_patterns_.end());
-
+  if (it == profiles_no_discard_patterns_.end()) {
+    // There's can be narrow window between profile creation and when prefs are
+    // read, which is when `profiles_no_discard_patterns_` is populated. During
+    // that time assume that a page might be opted out of discarding.
+    return true;
+  }
   return !it->second->MatchURL(url).empty();
 }
 
